registry/handlers/referrers.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/distribution/distribution/v3/manifest/ociartifact"
+	"github.com/distribution/distribution/v3/registry/api/errcode"
+	v2 "github.com/distribution/distribution/v3/registry/api/v2"
+	"github.com/distribution/distribution/v3/registry/storage"
+	"github.com/distribution/distribution/v3/registry/storage/driver"
+	"github.com/gorilla/handlers"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/specs-go"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// referrersDispatcher takes the request context and builds the
+// appropriate handler for handling referrers requests.
+func referrersDispatcher(ctx *Context, r *http.Request) http.Handler {
+	dgst, err := getDigest(ctx)
+	if err != nil {
+		if err == errDigestNotAvailable {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				ctx.Errors = append(ctx.Errors, v2.ErrorCodeDigestInvalid.WithDetail(err))
+			})
+		}
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx.Errors = append(ctx.Errors, v2.ErrorCodeDigestInvalid.WithDetail(err))
+		})
+	}
+
+	referrersHandler := &referrersHandler{
+		Context: ctx,
+		Digest:  dgst,
+	}
+	return handlers.MethodHandler{
+		"GET": http.HandlerFunc(referrersHandler.GetReferrers),
+	}
+}
+
+// referrersHandler handles http operations on referrers.
+type referrersHandler struct {
+	*Context
+	Digest digest.Digest
+}
+
+// GetReferrers fetches the list of referrers as an image index from the storage.
+func (h *referrersHandler) GetReferrers(w http.ResponseWriter, r *http.Request) {
+	dcontext.GetLogger(h).Debug("GetReferrers")
+
+	if h.Digest == "" {
+		h.Errors = append(h.Errors, v2.ErrorCodeManifestUnknown.WithDetail("digest not specified"))
+		return
+	}
+
+	annotations := map[string]string{}
+	query := r.URL.Query()
+	artifactTypeFilter := query.Get("artifactType")
+	if artifactTypeFilter != "" {
+		annotations[v1.AnnotationReferrersFiltersApplied] = "artifactType"
+	}
+	referrers, err := h.generateReferrersList(h, h.Digest, artifactTypeFilter)
+	if err != nil {
+		if _, ok := err.(distribution.ErrManifestUnknownRevision); ok {
+			h.Errors = append(h.Errors, v2.ErrorCodeManifestUnknown.WithDetail(err))
+		} else {
+			h.Errors = append(h.Errors, errcode.ErrorCodeUnknown.WithDetail(err))
+		}
+		return
+	}
+
+	response := v1.Index{
+		Versioned:   specs.Versioned{SchemaVersion: 2},
+		MediaType:   v1.MediaTypeImageIndex,
+		Manifests:   referrers,
+		Annotations: annotations,
+	}
+
+	w.Header().Set("Content-Type", v1.MediaTypeImageIndex)
+	enc := json.NewEncoder(w)
+	if err = enc.Encode(response); err != nil {
+		h.Errors = append(h.Errors, errcode.ErrorCodeUnknown.WithDetail(err))
+		return
+	}
+}
+
+func (h *referrersHandler) generateReferrersList(ctx context.Context, subjectDigest digest.Digest, artifactType string) ([]v1.Descriptor, error) {
+	dcontext.GetLogger(ctx).Debug("(*referrersHandler).generateReferrersList")
+	var referrers []v1.Descriptor
```

Comment:
nit: code style: let's only declare variables when needed.
suggestion: move this line after line 99 but before line 102

Reply:
moved accordingly.